causallearn/utils/KCI/KCI.py

```diff
@@ -339,7 +339,7 @@ def kernel_matrix(self, data_x, data_y, data_z):
                 if self.est_width == 'median':
                     kernelX.set_width_median(data_x)
                 elif self.est_width == 'empirical':
-                    kernelX.set_width_empirical_kci(data_x)
+                    kernelX.set_width_empirical_kci(data_z)
```

Comment:
Oh I'm not sure: you mean `kernelX`'s and `kernelY`'s width should be set empirically both from `data_z`, instead of `data_x` and `data_y` themselves?

Reply:
But this doesn't matter for empirical width right? `data_z` and `data_x`, `data_y` have the same sample size?

Reply:
Yes, according to the Matlab version, the empirical kernel width of `kernelX` and  `kernelY` should depend on the shape of Z, not X or Y.

Reply:
Yes, they have the same sample size but a different column size. Here, `data_x = np.concatenate((data_x, 0.5 * data_z), axis=1)`, so `data_x.shape[1]` is 2, while `data_z.shape[1]` is 1.

Reply:
I see! Wow, thanks for being so careful!!

Then,
+ Just for clarity, could you please add a comment and reference to the original code? (so that readers won't feel like it's a typo).
+ Is there any empirical performance gain / theoretical guarantee for this change? (so that we can ensure that it's not a typo in the matlab version).
+ After this change, will KCI in python and in matlab return the same value over the same dataset (try to fix randomness)?
+ And does this have anything to do with https://github.com/cmu-phil/causal-learn/issues/56?

Reply:
@aoqiz Could you address Haoyue's question? I am also curious....

Reply:
Sure, I'm working on that.

Reply:
@aoqiz Could you answer @MarkDana's question? I am also curious haha

1. "Is there any empirical performance gain / theoretical guarantee for this change? (so that we can ensure that it's not a typo in the matlab version)." I am also worried about this, and I don't know KCI much. To just make sure, in the paper, here we should data_z shape, right? And it's theoretically sound?

2. based on your updated description, if I understand correctly, now we can ensure almost identical result between matlab and python now, right?

3. Could you take a look at the issue @MarkDana mentioned? I am also curious whether it's related

Reply:
@tofuwen @MarkDana I'm sorry, I think I cannot answer these questions now, and I'm still investigating. I hope I can give you an answer before this weekend. Thanks for your patience!

> Is there any empirical performance gain / theoretical guarantee for this change? (so that we can ensure that it's not a typo in the Matlab version).

Actually, I didn't find anything about this in the paper, and I'm discussing it with Mingming now.

> After this change, will KCI in python and Matlab return the same value over the same dataset (try to fix randomness)?

After fixing that bug, they can return almost the same p-value over the same dataset. For three datasets, the program was run 100 times on each dataset, the mean of the p-value returned by Python and Matlab are 0.2494 and 0.2463 respectively (unconditional KCI); 0.36024, 0.3585 (conditional KCI); 0.34238, 0.3381 (unconditional KCI); 0.4971, 0.4910 (conditional KCI); 0.62922, 0.6281 (unconditional KCI); 0.3795, 0.3793 (conditional KCI). 

But I think the mean of the p-value should be the same for Python and Matlab if it's only due to the randomness. So what I'm doing now is checking whether there is any difference in the deterministic part.

> Could you take a look at the issue @MarkDana mentioned? I am also curious whether it's related

Once I fix the difference in the deterministic part, I'll look into this issue.

Thanks for your patience again!